File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    bool isPalindrome(string &mine) {
        int i = 0, j = mine.size() - 1;
        while (i <= j) {
            if (mine[i] != mine[j]) return false; 
            i++;
            j--;
        }
        return true; 
    }
    void loki (string s, int idx, vector<vector<string>>& ans, vector<string>& q){
        if(idx == s.length()){
            ans.push_back(q);
            return;
        } 
        for(int i = idx; i<s.length(); i++){
            string temp = s.substr(idx, i-idx+1);
            if(isPalindrome(temp)){
                 q.push_back(temp);
                 loki(s, i+1, ans, q);
                 q.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> ans;
        vector<string> q;
        loki(s, 0, ans, q);
        return ans;
    }
};

File: 131-palindrome-partitioning/palindrome-partitioning.py
class Solution:

    def partition(self, s: str) -> List[List[str]]:
        ans = []
        path = []
        def isPalindrome(l, r):
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True
        def dfs(idx):
            if idx == len(s):
                ans.append(path.copy())
                return 
            for end in range(idx, len(s)):
                if isPalindrome(idx, end):
                    path.append(s[idx:end+1])
                    dfs(end+1)
                    path.pop()
        dfs(0)
        return ans 
